asdf/extension/_compressor.py

```diff
@@ -0,0 +1,84 @@
+"""
+Compressor is an interface to implement extensions to the compression
+module. Extensions will typically subclass the Compressor ABC and
+provide that subclass as a setuptools entry point.  Decompressor
+is likewise used for decompression.
+
+Note that this interface has similar patterns to Converter.  This
+interface is designed for compression and decompression of ASDF
+binary array blocks, while Converter is designed for serialization
+of custom Python types into the YAML tree.
+"""
+
+import abc
+
+class Compressor(abc.ABC):
+    """
+    Abstract base class for plugins that compress binary data.
+
+    Implementing classes must provide the `labels` property, and
+    the `compress()` method.
+    May also provide a constructor that takes kwargs that are
+    set via `asdf.get_config().compression_options[label] = kwargs`.
+    """
+    @classmethod
+    def __subclasshook__(cls, C):
+        if cls is Compressor:
+            return (hasattr(C, "labels") and
+                    hasattr(C, "compress"))
+        return NotImplemented # pragma: no cover
+
+
+    @abc.abstractproperty
+    def labels(self):
```

Comment:
Do you have a use case in mind for multiple labels?  If that's uncommon, it might be reasonable to force users to create one instance of the class per label.

Reply:
Speaking of instances, did you have in mind that the `Extension` instance would return instances of `Compressor` subclasses or the subclasses themselves?  The `Extension.converters` property returns instances and I prefer that they work the same way if possible.  Do you foresee a need for `Compressor` instances that maintain some kind of internal state between calls to `compress` or `decompress`?

Reply:
My thought was for label aliases.  I think @perrygreenfield had mentioned such aliasing, but that was probably at the YAML level (or maybe in the new JSON block header, if that's still the plan).  Otherwise, I agree restricting a `Compressor` to one label might be preferable.

I think I'm okay with the `Extension` returning instances, although the `Compressor` objects are currently designed to take arguments from the config in their constructors.  So that logic would have to be moved to the `Extension` and either implemented by the user writing the extension, or wrapped in some way.  I don't foresee any particular need to maintain state between calls, except perhaps for tracking performance metrics, which can be done by other means.